### PR TITLE
fix: support for ghc 9.6 - Remove MonadTrans

### PR DESCRIPTION
**Pull request by guibou**
`Monad` for `TardisT` requires that the underlying `Monad` be an instance of `MonadFix`, which is not general enough for `TardisT` to be an instance of `MonadTrans` since `MonadTrans` got a `forall m. Monad m => Monad (t m)` constraint. (where `t ~ TardisT a b` in our context) in transformers `0.6`.

This commit removes the `MonadTrans` instance and instead provides a `liftTardis` function which will hopefully cover most use case previously covered by `lift`.

See https://github.com/DanBurton/tardis/issues/15 for discussion.

### CODE REVIEW

**Comment by guibou** [comment]
Note: I'm opening this MR for reference and so user of GHC 9.6 can update `tardis` without too much pain. I'm unsure that this is the right solution or if we need to do more refactoring.

File: src/Control/Monad/Trans/Tardis.hs [hunk 1]
```diff
@@ -53,6 +54,8 @@ import Control.Monad.Morph
 -- a 'backwards-traveling' state and a 'forwards-traveling' state.
 -- This library consistently puts the backwards-traveling state first
 -- whenever the two are seen together.
+--
+-- Note that even if this library follows the definition of a monad transformer, as described in 'MonadTrans', this is not a true monad transformer, because the underlying 'Monad' needs to be an instance of 'MonadFix', which is not general enough for 'MonadTrans'.
```
Comment (Lysxia) [inline src/Control/Monad/Trans/Tardis.hs:58]: What definition are you referring to? "A monad transformer makes a new monad out of an existing monad" reads to me like `forall m. Monad m => Monad (t m)` which is what `TardisT` doesn't satisfy.

**Comment by guibou** [comment]
Thank to @Lysxia, `MonadTrans` is back, but now that's `MFunctor` which is not.

**Comment by guibou** [comment]
@DanBurton what should we do with this MR?

You pushed https://github.com/DanBurton/tardis/commit/e0a9b2e019630c717ce87ed3ed2218b74091f052 and released a new version of tardis which removes the `MonadTrans` instance. On the other hand, this MR removes the `MFunctor` instance and imposes a new `MonadFix` constraint on the inner monad in `TardisT`.

To be honest, I was surprised by the update of tardis 0.5 and when trying to apply it to my codebase (which currently uses this MR), it breaks (because I rely on the `MonadTrans` instance).

**Comment by guibou** [comment]
I fixed `MFunctor`, but I have no idea if that's correct.

**Comment by guibou** [comment]
I'm unsure the `MFunctor` thing is correct, and I don't know how to fix it for good.

I'm closing. Maybe I'll find a way to not use `MonadTrans` in my codebase, or I'll use a fork of `tardis`.

Thank you for maintaining `tardis` and I do apologize, my latest messages were rough.

**Comment by DanBurton** [comment]
@guibou Sorry I forgot to respond earlier. You should be able to use `liftTardisT` in places where you previously used `lift`. I'd rather keep the Tardis representation as is and acknowledge that it is no longer a `MonadTrans` according to the new constraints. Closing the PR is the right move now, but I appreciate the effort you've gone to in making it nonetheless.

**Pull request closed without merging**